ui/pages/confirmations/utils/token.ts

```diff
@@ -1,32 +1,89 @@
 import { memoize } from 'lodash';
 import { Hex } from '@metamask/utils';
+import { AssetsContractController } from '@metamask/assets-controllers';
 import { getTokenStandardAndDetails } from '../../../store/actions';
 
+export type TokenDetailsERC20 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC20Standard']>['getDetails']
+  >
+> & { decimalsNumber: number };
+
+export type TokenDetailsERC721 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC721Standard']>['getDetails']
+  >
+>;
+
+export type TokenDetailsERC1155 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC1155Standard']>['getDetails']
+  >
+>;
+
+export type TokenDetails =
+  | TokenDetailsERC20
+  | TokenDetailsERC721
+  | TokenDetailsERC1155;
+
 export const ERC20_DEFAULT_DECIMALS = 18;
 
-/**
- * Fetches the decimals for the given token address.
- *
- * @param {Hex | string} address - The ethereum token contract address. It is expected to be in hex format.
- * We currently accept strings since we have a patch that accepts a custom string
- * {@see .yarn/patches/@metamask-eth-json-rpc-middleware-npm-14.0.1-b6c2ccbe8c.patch}
- */
-export const fetchErc20Decimals = memoize(
-  async (address: Hex | string): Promise<number> => {
+export const parseTokenDetailDecimals = (
+  decStr?: string,
+): number | undefined => {
+  if (!decStr) {
+    return undefined;
+  }
+
+  for (const radix of [10, 16]) {
+    const parsedDec = parseInt(decStr, radix);
+    if (isFinite(parsedDec)) {
+      return parsedDec;
+    }
+  }
+  return undefined;
+};
+
+export const memoizedGetTokenStandardAndDetails = memoize(
+  async (
+    tokenAddress?: Hex | string,
+    userAddress?: string,
+    tokenId?: string,
+  ): Promise<TokenDetails | Record<string, never>> => {
     try {
-      const { decimals: decStr } = await getTokenStandardAndDetails(address);
-      if (!decStr) {
-        return ERC20_DEFAULT_DECIMALS;
-      }
-      for (const radix of [10, 16]) {
-        const parsedDec = parseInt(decStr, radix);
-        if (isFinite(parsedDec)) {
-          return parsedDec;
-        }
+      if (!tokenAddress) {
+        return {};
       }
-      return ERC20_DEFAULT_DECIMALS;
+
+      return (await getTokenStandardAndDetails(
+        tokenAddress,
+        userAddress,
+        tokenId,
+      )) as TokenDetails;
     } catch {
-      return ERC20_DEFAULT_DECIMALS;
+      return {};
```

Comment:
Rather than returning an empty object, shouldn't we just let it throw and return `ERC20_DEFAULT_DECIMALS` on `catch`? In that way we could eliminate optional parameter in `parseTokenDetailDecimals` and also early return?

Reply:
We want to default decimals only for ERC20 tokens.

Reply:
We need to know when we are using default value in `useGetTokenStandardAndDetails` so that we can capture this in metrics.

Reply:
I would avoid re-throwing the error as it is existing behaviour and this method is used at few places.

Reply:
> We need to know when we are using default value in useGetTokenStandardAndDetails so that we can capture this in metrics.

We can use `error` property in the `useGetTokenStandardAndDetails` if it throws

Reply:
`useGetTokenStandardAndDetails ` is using `memoizedGetTokenStandardAndDetails`. This change will require change in that hook also and handling error wherever we use `memoizedGetTokenStandardAndDetails`.